gdx/src/com/badlogic/gdx/graphics/g2d/PixmapPacker.java

```diff
@@ -326,66 +322,61 @@ public synchronized void dispose () {
+	 */
+	public synchronized void updateTextureAtlas(TextureAtlas atlas, TextureFilter minFilter, TextureFilter magFilter,
 		boolean useMipMaps) {
 		for (Page page : pages) {
-			if (page.texture == null) {
-				if (page.rects.size != 0 && page.addedRects.size > 0) {
-					page.texture = new Texture(new PixmapTextureData(page.image, page.image.getFormat(), useMipMaps, false, true)) {
-						@Override
-						public void dispose () {
-							super.dispose();
-							getTextureData().consumePixmap().dispose();
-						}
-					};
-					page.texture.setFilter(minFilter, magFilter);
-
-					for (String name : page.addedRects) {
-						Rectangle rect = page.rects.get(name);
-						TextureRegion region = new TextureRegion(page.texture, (int)rect.x, (int)rect.y, (int)rect.width,
-							(int)rect.height);
-						atlas.addRegion(name, region);
-					}
-					page.addedRects.clear();
-				}
-			} else {
-				if (page.addedRects.size > 0) {
+			if (page.addedRects.size > 0) {
+				if (page.texture == null) {
+					generatePageTexture(page, minFilter, magFilter, useMipMaps);
+				} else {
 					page.texture.load(page.texture.getTextureData());
-					for (String name : page.addedRects) {
-						Rectangle rect = page.rects.get(name);
-						TextureRegion region = new TextureRegion(page.texture, (int)rect.x, (int)rect.y, (int)rect.width,
+				}
+				for (String name : page.addedRects) {
+					Rectangle rect = page.rects.get(name);
+					TextureRegion region = new TextureRegion(page.texture, (int)rect.x, (int)rect.y, (int)rect.width,
 							(int)rect.height);
-						atlas.addRegion(name, region);
-					}
-					page.addedRects.clear();
+					atlas.addRegion(name, region);
 				}
+				page.addedRects.clear();
+			}
+		}
+	}
+
+	private void generatePageTexture(final Page page, TextureFilter minFilter, TextureFilter magFilter, boolean useMipMaps) {
+		page.texture = new Texture(new PixmapTextureData(page.image, page.image.getFormat(), useMipMaps, false, true)) {
+			@Override
+			public void dispose() {
+				super.dispose();
+				getTextureData().consumePixmap().dispose();
+			}
+		};
+		page.texture.setFilter(minFilter, magFilter);
+	}
+
+	/** Return an ordered array of texture regions corresponding with the pages of the packer, creating new textures
+	 * if necessary.  The same array will be returned from successive calls to this method, possibly modified with
+	 * additional texture regions if more pages have been added since the last call.  Once this method has been called,
+	 * you must dispose of the resulting texture regions (or equivalently, any atlas generated or updated by this packer)
+	 * yourself - disposing the PixmapPacker will no longer dispose these textures or their backing pixmaps.
+	 */
+	public Array<TextureRegion> getTextureRegions(TextureFilter minFilter, TextureFilter magFilter, boolean useMipMaps) {
+		while (textureRegions.size < pages.size) {
+			if (disposed) return null;

```

Comment:
I think it's better to pass in an `Array<TextureRegion>`. Really, PixmapPacker should only be packing rects. It has its one job, to pack Pixmaps. It shouldn't know about TextureRegion or TextureAtlas. We only jam on some methods for those for convenience, because we don't really have a better place to put them, and they are discoverable there. PixmapPacker should avoid holding state that isn't really its business. Eg, maybe I pack to pixmaps and write them to disk, not using TextureRegions at all.

I wonder if the old dispose was better. If you call it, it disposes all the pixmaps with no checking the texture field. This is clear what happens. If you are using `updateTextureAtlas` or `updateTextureRegions`, just don't dispose the packer so the pixmaps live until the atlas or regions are disposed. I guess the new dispose is safer, though it may not dispose everything.